### ISSUE TITLE
Pass through data- attributes

### ISSUE THREAD

**Issue by mojavelinux** [asciidoctor/asciidoctor#1305]
Pass through attributes on blocks that begin with `data-` to the HTML output. This will allow for tighter integration with dynamic pages that rely on this information.

**Comment by viebel** [comment]
That would be great!

**Comment by rmannibucau** [comment]
+1, need that as well

**Comment by aggsol** [comment]
+1 I also would like to use this feature!

**Comment by aggsol** [comment]
This feature would really help me. I will to set a bounty of $50 to this feature. If you do not accept bounties I am willing to donate to the [Eclipse Foundation](https://www.eclipse.org/donate/) or the [Free Software Foundation](https://www.fsf.org/about/ways-to-donate/).

**Comment by mojavelinux** [comment]
We use currently use BountySource to manage bounties. https://www.bountysource.com/issues/10270083-pass-through-data-attributes I ensure that whoever works on the issue receives their owed share of the award.

**Comment by aggsol** [comment]
> We use currently use BountySource to manage bounties. https://www.bountysource.com/issues/10270083-pass-through-data-attributes I ensure that whoever works on the issue receives their owed share of the award.

Ok, I posted a bounty over there.

**Comment by UnicodingUnicorn** [comment]
Sorry, I am fairly new to asciidoc. I just want to confirm that if you have something like:

```
:data-foo1: bar
== Section

:data-foo2: bar
stuff
```

`data-foo1` and `data-foo2` along with their values will get passed along as attributes to the top most HTML element of their respective blocks.

Also, are there such a thing as attribute definitions for inline elements like links? My reading of the docs leans on no, but still inconclusive.

**Comment by aggsol** [comment]
@UnicodingUnicorn  The attributes are passed along as classes to the HTML

```asciidoc
=== 5
[.progress=23]
Lorem ipsum dolor sit
```
The result looks like this

```html
<div class="paragraph progress=23">
<p>Lorem ipsum dolor sit</p>
</div>
```

**Comment by briandominick** [comment]
Yes we are not talking about AsciiDoc attributes but rather HTML element attributes (tag arguments). So as of now, @UnicodingUnicorn has no way to pass those data attributes, but the method of defining AsciiDoc attributes named as `data-*` inline will not be the solution here.

Right now, we have two ways each to pass an id or class attribute to any HTML that results from an AsciiDoc block, and most specialized blocks carry some relevant attribute info through the conversion. All we are really lacking is a means of arbitrarily naming and assigning values to [HTML data attributes](https://www.w3schools.com/tags/att_data-.asp) used mainly for letting HTML elements interact with JavaScript after rendering.

So we want the following two AsciiDoc snips:

`[#thing,role='some',data-foo='bar',data-foo-two='bar2']`

`[id='thing',.some,data-foo='bar',data-foo-two='bar2']`

Both of these should evaluate to...

`<element id="thing" class="some" data-foo="bar" data-foo-two="bar2">`

Here is what [HTML5 spec](https://www.w3.org/TR/html50/dom.html#embedding-custom-non-visible-data-with-the-data-*-attributes) says about data attribute format:
> A custom data attribute is an attribute in no namespace whose name starts with the string "data-", has at least one character after the hyphen, is XML-compatible, and contains no uppercase ASCII letters.

I'm considering taking this on. Am I on the right track? What else does this call for? Are there other ways we indicate this stuff?

**Comment by UnicodingUnicorn** [comment]
Yes, that was what I was thinking, too. I just wanted to verify how attributes are passed in asciidoc so I don't end up barking up the wrong tree, since there appear to be two ways: `:foo:` above the block and `[foo=bar]` inline.

I'm already halfway done with my implementation, but if you want to continue this, do feel free.

**Comment by briandominick** [comment]
Where did you get the "`:foo:` above the block" option? I hadn't seen that documented, and I'm unable to get it to work. It doesn't seem like a bad idea, especially if we're going to need the attribute for substitution later in the doc. But I think the real problem is we're not trying to add arbitrary attributes, just data attributes. Since there is no HTML5 attribute called *foo*, we aren't trying to enable passing just anything, IMO. We want to restrict it to the spec's allowance for `data-*` attributes, specifically. But do let me know if I'm missing something.

**Comment by UnicodingUnicorn** [comment]
Sorry, I was just using a `foo` as a placeholder for anything, should have written `data-*`. I've seen putting `:attribute-name: attribute-value` mentioned [here](https://www.methods.co.nz/asciidoc/chunked/ch28.html), and it works nicely with asciidoctor. I've already done the code to pass through attributes defined as such.

What I was trying to clarify was if attributes defined like `[data-foo="bar"]` actually count as attributes. Currently I am working on the code to extract them from the class string, where they are currently passed to.

**Comment by briandominick** [comment]
Okay, it's really possible that I'm missing something big (and have been for years now), but I _think_ our disconnect here is that you're using the term _attribute_ interchangeably, and there are two major intended uses here. I cannot see anywhere that *HTML attributes* can be set simply be establishing *AsciiDoc attributes* in ahead of a new section.

AsciiDoc attributes are properties and variables that get carried within an AsciiDoc document.

HTML attributes are the properties carried _within_ an HTML element.

I will say I like your idea of setting HTML attributes along with AsciiDoc attributes this way, but unfortunately I think it will be backward incompatible if it was never defined previously. If someone has ever set something like this:

```
:data-name: something
== Some Section
```

We will suddenly be passing `data-name` attribute and value into the `<h2 data-name="something">` or whatever, which doesn't seem super dangerous but is not something they could have expected. Unless I am dead wrong about this and missing something.

So I think all we should be trying to do with this change is add the ability to pass a `data-*` HTML attribute to a block being defined with headers like so:

```
[data-name='something']
== Some Section
```

**Comment by briandominick** [comment]
Also a side note: I think if you find something in the legacy Python AsciiDoc docs that has somehow not been honored/implemented in Asciidoctor AsciiDoc, it is good to raise them here and get them implemented. But for the most part, I think standardizing around concepts described in the Asciidoctor.org docs is better than pointing to explanations in the Python docs, which most of us are less familiar with and (I'd argue) find far less useful.

**Comment by UnicodingUnicorn** [comment]
Ah I see, I interpreted the wording of this issue as 'asciidoc attribute => html attribute if asciidoc attribute begins with `data-`'. Sorry, I think I misread things.

**Comment by briandominick** [comment]
No I'm really glad we cleared it up. I absolutely _hate_ that we call AsciiDoc attributes _attributes_, rather than parameters, properties, variables, or just about anything else... but here we are.

I may take a swing at this on the weekend if you have not. If you are going for the $50, I'm happy to step aside or just help out, since I already donate to Asciidoctor (and make my living off it).

**Comment by mojavelinux** [comment]
@briandominick Yes, your proposal is the direction we have in mind. A *block* attribute that begins with `data-` would be mapped directly to an HTML `data-` attribute. The idea being that it should be passed on directly to the output.

So, for example, if we had the following paragraph:

```
[data-name=value]
paragraph
```

That would produce the following HTML:

```
<div class="paragraph" data-name="value">
<p>paragraph</p>
</div>
```

Asciidoctor extensions could also access these attributes on the block, so they might be useful during processing as well.

**Comment by mojavelinux** [comment]
> since there appear to be two ways: :foo: above the block and [foo=bar] inline.

These are two very different things. The first is a document attribute, and it's always associated with the document. The latter is a block attribute, and it gets associated with the block below it.

A document attribute is intended for a) storing reusable content and b) configuring the behavior of the processor. It doesn't have any relation to blocks, other than the fact that content in those blocks that follow the declaration of the document attribute can reference its value.

The `data-` attributes only pertain to block attributes. Document attributes do not have a role to play here.

**Comment by mojavelinux** [comment]
> So I think all we should be trying to do with this change is add the ability to pass a data-* HTML attribute to a block being defined with headers like so:

Correct. To rephrase, we would be adding the ability to define a `data-` attribute in HTML by defining a block attribute with the same name in AsciiDoc. This mapping would be scoped to the block.

**Comment by mojavelinux** [comment]
> I absolutely hate that we call AsciiDoc attributes attributes, rather than parameters, properties, variables, or just about anything else... but here we are.

I don't think the terminology is problematic. Attribute is information "attributed" to something else. Here, we're augmenting either the document (a document attribute), the block (a block attribute), or the inline node (an inline attribute) with additional information (i.e., metadata). This is exactly the same usage of the terminology as in HTML.

Document attributes go a little further than how it's defined in HTML since they can also be referenced by the content, so they act like variables. This is similar to CSS's `attr()` function, except it can be applied more broadly.

So while attributes can be variables, can be parameters, can be properties, they can be any of these things and hence the attribute terminology applies.

...of course, this is a conversation we can have in the AsciiDoc spec, when we need to standardize the terminology.

**Comment by mojavelinux** [comment]
> I may take a swing at this on the weekend if you have not. 

My one reservation is that I really want to hold off adding this feature to the existing HTML converter and instead make it part of the new, modern HTML converter. For one, the existing converter doesn't delegate HTML attribute building to a common method, so it would require adding duplicate logic to every node type. But it's also because we don't want to keep investing in the existing HTML converter. We want to put that time into the new one. See #242.

If you really wanted this today, you could extend the HTML5 converter and add data attributes in strategic locations (such as on paragraphs). There's nothing you have add to the parser. This all happens in the converter. For example:

```
class Html5ConverterWithDataAttributes < (Asciidoctor::Converter.for 'html5')
  register_for 'html5'
  def convert_paragraph node
    # build HTML for paragraph including data- attributes
  end
end
```

**Comment by briandominick** [comment]
> I don't think the terminology is problematic.

I've never heard of "attributes" used at the document level. I think even DocBook only uses it at the block level. Adding in the weird use of _attributes_ used in place of _variables_ for arbitrary substitutions -- that one hangs up nearly everybody I try to teach AsciiDoc to, whether they have a dev background or not. It's very unnatural and threw me for a long time, and right here we have a great example of it being super confusing.

Your case for why "attributes" is technically _accurate_ and _appropriate_ was a good read, and of course you're right. Then again, _usability_ has little to do with either accuracy or appropriateness. A term that does not conflict with other concepts seems highly preferable, but _attripropervariameters_  is a fraught space for finding such clarity of terms. That said, I'm not even sure it would be worth trying to change for the spec, but it might be worth discussing. _Propervariabutes_? I'll keep trying...

> My one reservation is that I really want to hold off adding this feature to the existing HTML converter and instead make it part of the new, modern HTML converter.

This makes a lot of sense. I do think for most use cases this feature has a viable workaround in manipulating role/class designations such that they can effectively pass data to the front end, but I have run into this several times and would love to deliver it to users. I even found myself using HTML where I might have used AsciiDoc recently because of this limitation. The horror. Nevertheless, i may dig for another issue to work on or just get back to my own stuff.

**Comment by gdams** [comment]
For anyone looking here, I did manage to find a workaround (although it feels like a bit of a hack). I wanted to get tooltips working with boostrap so used the following code:

```adoc
Linux (x64) ["data-bs-toggle="tooltip"data-bs-placement="right"title="Linux (x64) builds should work on any distribution with glibc version 2.12 or higher."]#^[1]^#
```

which in turn rendered the following `span`:

```html
<span class="" data-bs-toggle="tooltip" data-bs-placement="right" title="Linux (x64) builds should work on any distribution with glibc version 2.12 or higher." "=""><sup>[1]</sup></span>
```

**Comment by Bonjour123** [comment]
I don't understand, what is the current advancement for this feature ? It seems that every one here agrees that it's technically simple to do, so I would think it's finished by 2022.. Am I missing something ? Because I couldn't find information in the docs and I tried all combinations, but couldn't achieve to pass through data-* attributes

**Comment by Bonjour123** [comment]
So I did what you suggested in [here](https://github.com/asciidoctor/asciidoctor/issues/1305#issuecomment-521851054)
I downloaded the IntelliJ AsciiDoc Plugin from [here](https://plugins.jetbrains.com/plugin/7391-asciidoc)
Then I modified the \lib\asciidoctorj-2.5.6.jar\gems\asciidoctor-2.0.17\lib\asciidoctor\converter\html5.rb file doing the following:
I replaced 

      def convert_paragraph node
          if node.role
            attributes = %(#{node.id ? %[ id="#{node.id}"] : ''} class="paragraph #{node.role}")
          elsif node.id
            attributes = %( id="#{node.id}" class="paragraph")
          else
            attributes = ' class="paragraph"'
          end

with 

    def convert_paragraph node
        if node.role
          attributes = %(#{node.id ? %[ id="#{node.id}"] : ''} class="paragraph #{node.role}")
        elsif node.id
          attributes = %( id="#{node.id}" class="paragraph")
        else
          attributes = ' class="paragraph"'
        end    
        if node.tags
          attributes += ' tags=#{node.tags}'
        end

Now I get the following error:
      unable to render AsciiDoc document
      
      org.asciidoctor.jruby.internal.AsciidoctorCoreException: org.jruby.exceptions.NoMethodError: (NoMethodError) undefined method `tags' for #<Asciidoctor::Block:0x1ed003b1>
      
      org.jruby.exceptions.NoMethodError: (NoMethodError) undefined method `tags' for #<Asciidoctor::Block:0x1ed003b1>

I guess the parser needs to pass the "tags" attribute down to the converter but you wrote that nothing has to be changed in the parser, so I don't understand what is missing. 
I'm sure it's not a big deal, but I've never touched Ruby, so if you could help me on this one, it would be really great !

**Comment by mojavelinux** [comment]
Please ask usage questions in the project chat at https://chat.asciidoctor.org. I don't want issues turning into support threads. Thanks.

**Comment by mojavelinux** [comment]
FYI, this has been implemented in the reveal.js converter, and will likely be included as a stated expectation in the AsciiDoc Language specification.

**Comment by mojavelinux** [comment]
As I suggested it would be, this is now included in the AsciiDoc Language specification outline.

This feature will be introduced into Asciidoctor in the next generation HTML converter. You can follow it's progress at https://github.com/asciidoctor/asciidoctor/projects/1?query=is%3Aopen+sort%3Aupdated-desc. I don't have plans at this time to add it to the current HTML converter.

With that said, there is no reason to have to wait. You can create your own extended converter or converter templates and pass these attributes through to the HTML. That's what the reveal.js converter does.

**Comment by mojavelinux** [comment]
If you need help writing a converter, please feel free to ask in the project chat at https://chat.asciidoctor.org.
